src/Plugin.js

```diff
@@ -32,49 +33,63 @@ class Plugin {
     }
 
     /**
-     * Initialize the current plugin,
+     * Initialize the current plugin, if necessary.
+     *
      * @abstract
      */
     init() {}
 
     /**
-     * Return the type of this plugin. This can be one of the
-     * following:
-     *
-     * <ul>
-     * <li>rule - this plugin implements a new rules
-     * <li>parser - this plugin knows how to parse files more deeply
-     * than line-by-line
-     * <li>formatter - this plugin formats results for a particular
-     * type of output
-     * </ul>
+     * For a plugin that implements rules, this returns a list of Rule
+     * classes that this plugin implements. Note this is the class itself,
+     * not an instance of the class. The linter may need to instantiate
+     * this rule multiple times with different optional parameters.
      *
-     * @returns {String} tells what type of plugin this is
-     * @abstract
+     * @returns {Array.<Class>} list of Rule classes implemented by this
+     * plugin
      */
-    getType() {
-        return this.type;
+    getRules() {
+        return [];
     }
 
     /**
-     * For a "rule" type of plugin, this returns a list of Rule instances
-     * that this plugin implements.
+     * Return a number of pre-defined rule sets. The idea behind this
+     * method is that the plugin can define sets of rules that users of
+     * the plugin can rely on. As the plugin
+     * developer adds new rules in their plugin, they can also update
+     * the rule set to include those new rules and users of this plugin
+     * will get enhanced functionality automatically without changing
+     * their own configuration.
      *
-     * @returns {Array.<Rule>} list of Rule instances implemented by this
-     * plugin
+     * For example, if there is a plugin named
+     * "android", the plugin writer can add support for Java, Kotlin,
+     * and properties files in the same plugin by adding parsers and rules
+     * for each file type. They can then also add rulesets called "java",
+     * "kotlin" and "properties" which will apply all the rules from this
+     * plugin that are appropriate for the file types.
+     *
+     * By convention, these rulesets are named the same as the file type
+     * that they support, but this is not a strict requirement. Plugin
+     * writers should document the rulesets that the plugin supports in
+     * the README.md for that plugin so that users know that it is available.
+     *
+     * @returns {Object} an object where the properties are the names of
+     * rulesets and the values are objects that configure a ruleset. The
+     * properties of this subobject are the names of the rules and the
+     * values are the optional parameters for the rule, or "true" to indicate
+     * that the rule should be turned on for this set.
      */
-    getRules() {
-        return [];
+    getRuleSets() {
```

Comment:
Question) Don't we need to have `getRuleSets()` as well?

Reply:
There is a method for getRules and getRuleSets. Also, getType() was removed because it is in the Rule instead, so it makes the diff a little confusing.

Reply:
oh, sorry I was trying to ask setRuleSet(). not getRuleSet()

Reply:
The plugins only provide new rulesets, they don't manage them. The lint tool collects them and manages them in the RuleManager class.